Initial secret provision (#16)
ietf-wg-wimse/draft-ietf-wimse-workload-identity-practices: draft-ietf-wimse-workload-identity-practices.md

Accessing different protected resources may require steps 2) to 4) again with different scope parameters. Accessing a protected resource in an entirely different authorization domain often requires the entire flow to be followed again, to retrieve a new platform-issued credential with an audience for the external authorization server. This, however, differs based on the platform and implementation.

## Credential issuance

Credentials can be provisioned to the workload through different mechanisms which present different complexity and security risks. The following section highlights the pros and cons of most widespread solutions.

> ### Enviromental variable

Reviewer: ```suggestion
### Environment variable
```

Injecting the credentials into the enviromental variables allows for simple and fast deployments. Applications can directly access them through system-level mechanism, e.g., through the env command in linux. This flexibility, however, comes wiht security drawbacks.

> * 1) enviromental variables are static in nature and more dynamic solutions require to introduces higher complexitiy.

Reviewer: environmental
```suggestion
* 1) environmental variables are static in nature and more dynamic solutions require to introduces higher complexity.
```

> * 2) performing access control to enviromental variable is not trivial and it might also not reach the same security results.

Reviewer: ```suggestion
* 2) performing access control to environmental variable is not trivial and it might also not reach the same security results.
```

> * 3) enviromental information are often used for debugging purpose and can be printed at start-up or inside logs.

Reviewer: ```suggestion
* 3) environmental information are often used for debugging purpose and can be printed at start-up or inside logs.
```

> Leveraging enviromental variable to provide credentials should be favored when the provided secrets have a short-term validity, i.e., an initial secret during the set-up of the application, and/or when simplicity is required, e.g., during the development of PoC.

Reviewer: ```suggestion
Leveraging environmental variable to provide credentials should be favoured when the provided secrets have a short-term validity, i.e., an initial secret during the set-up of the application, and/or when simplicity is required, e.g., during the development of PoC.
```

> Leveraging enviromental variable to provide credentials should be favored when the provided secrets have a short-term validity, i.e., an initial secret during the set-up of the application, and/or when simplicity is required, e.g., during the development of PoC.

Author: I tried to rephrase it, I'll push it with the next commit

> ### Volumes

Reviewer 2: There is this existing pattern where a another process stores the credential as a file which the workload picks up. Or in general, the platform when bootstrapping the file system puts a credential into it.

I think that mounts are just a way to attach it, for our purposes it doesn't matter whether it's a mount, link, or anything else. For the purpose here what matters is that access is via the file system, Open, Close, etc. not exactly how the file got there.

> Volumes, e.g., projected volumes in Kubernetes, allow to inject credential to the container through the file-system. This solution offers higher degree of flexibility thanks to the possibility of dynamically rotate the secrets and perform access control on the injected file. This additional security comes with additional complexity.

Reviewer: ```suggestion
Volumes, e.g., projected volumes in Kubernetes, allows to inject credentials into the container through the file-system. Such solutions enable secret rotation, and access control on the injected secret with standard operating system measure for files.
```

> * 1) access control to the mounted volume needs to be correctly configure to limit access from unwanted applications. Solution such as DAC (uid and guid) or MAC (seLinux) are avaible in most Operating Systems.

Reviewer: ```suggestion
* 1) access control to the mounted volume needs to be correctly configure to limit access from unwanted applications. Solution such as DAC (uid and guid) or MAC (SELinux) are available in most Operating Systems.
```

> * 2) isolating the mounted volumes ,e.g., through namespaces, is required to avoid that a compromised application is able to escape to the main OS and ,for example, retrieve private information of other processes.

Reviewer: ```suggestion
* 2) isolating the mounted volumes from critical host OS paths and processes is required. E.g, on Linux this is achieved by utilising namespaces.
```

> * 2) isolating the mounted volumes ,e.g., through namespaces, is required to avoid that a compromised application is able to escape to the main OS and ,for example, retrieve private information of other processes.

Author: In my view most of this limitations and solution are applied in container environments, do you think we should remark it anyway?

> * 3) credentials rotation requires a monitoring solution to detect near-to-expiration secrets and subtitute them. This is usually embedded into orchestrators such as Kubernetes, but requires proper configuration.

Reviewer: ```suggestion
* 3) credentials rotation requires a monitoring solution to detect near-to-expiration secrets and substitute them. This is usually embedded into orchestrators such as Kubernetes, but requires proper configuration.
```

> * 3) credentials rotation requires a monitoring solution to detect near-to-expiration secrets and subtitute them. This is usually embedded into orchestrators such as Kubernetes, but requires proper configuration.

Reviewer 2: I see this a problem outside of provisioning. All of the options require it to keep the lifecycle small without impacting the workload.

Author: I highlighted this problem  here, because for environmental variables the update is quite painful and usually quite not used, while some options of the network section, e.g., spiffe, the workload is responsible to request the updated credentials. Do you think we should just remove it or put as conclusion/introduction for all of them?

> Leveraging volumes to provision secrets offers all the required security tools, at the cost of a more complex configuration. This approach should be favored when multiple process might require to access the same set of credentials.

Reviewer 2: > when multiple process might require to access the same set of credentials.

I don't see this as a factor here, this can be achieved with all options (env variables, network, etc.) To me the benefits here are:
- no runtime coupling/asynchronous provisioning. A side-process can start, provision in file and terminate. Upon expiry it would start again (some scheduling) and refresh. This is not easily achievable with the other options which either require a peer to answer requests or are limited due to the isolation of environment variables.
- common support to read files.

I believe it's also worth mentioning that the pattern has established to have an environmental variable containing a file path where the actual credential is stored.

> ### Network inferaces

Reviewer: ```suggestion
### Network interfaces
```

> ### Network inferaces

Reviewer 2: Not all the options below require a network stack (UDS). Either we split those into 2 separate sections (which will be difficult as SPIFFE can be TCP too) or use a more broader term.

Maybe "Local APIs" or something along those lines?

> Network interfaces rely on the network stack to communicate between the Host and the container application. Some examples of this interface are UNIX Domain Socket (spiffe), loopback interface, Magic (Link-Local) Address (AWS Metadata service). These solutions offer on-demand identity rotation and ensure security thanks to network isolation, but introduce new complexity.

Reviewer 2: > These solutions offer on-demand identity rotation.

I see multiple flavors at the `on-demand` part. Either these APIs "push" new credentials actively to the workload (require some notification mechanism or open channel) or the workload itself needs to have logic to "re-request".

Some ideas from the top of mind:
> Local APIs offer the capability to re-provision updated credentials. Communication between workload and API allows the workload to re-request a new credential (or a different one). Based on the technology it is even possible to pro-actively distribute new credentials to workloads (e.g. upon expiry, during a revocation event or a change in permissions.

> * 1) the appliaction is required to support different network stack in orded to allow portability among different solutions.

Reviewer 2: Hm. If I understand you correctly you want to highlight the decreased flexibility of changing the implementation?

How about something like `The benefit of a more structured delivery comes with less portability between different APIs.`?

> * 1) the appliaction is required to support different network stack in orded to allow portability among different solutions.

Reviewer: ```suggestion
* 1) the application is required to support different network stack in order to allow portability among different solutions.
```

* 2) additional latency may be introduced due to the request and additional operational overhead.

> 

Reviewer 2: * 3) credentials are only issued when requested reducing unnecessary credential issuance.
* 4) allows for narrowly-scoped & short-lived credentials. For instance one-time credentials or narrowly scoped ones such as JSON Web Tokens with specific audiences.

> Network inferfaces increase flexibility with respect to on-demand credentials at the cost of application complexity. To avoid this overhead, they should be used in conjuction with side-cars application and centrally manages identity solution.

Author: I agree that the statement could be confusing. My intention was to state that side-car can lift the requirements on the application to support different APIs, and after retrieving the credentials they can propagate them to the application through more standardize solution.

But I also support your conclusion, not sure I would repeat them also in the pointed list, I was trying to point out more the limitations there. WDYT?
